### PR TITLE
Fall back to general completions when semantic engine has no suggestions

### PR DESCRIPTION
**Pull request by puremourning**
TL;DR
=====

- supply general completions (alongside any errors raised) when the semantic completer (`clang_completer`, `jedi_completer`, `omnisharp_completer`, etc.) returns no results for the query and semantic completion is not forced
- tag results from the identifier completer with `[ID]`
- add tests for the above cases

Background
=========

Submitted in reference to the following issues:

* https://github.com/Valloric/YouCompleteMe/issues/1506
* https://github.com/Valloric/YouCompleteMe/issues/1225
* https://github.com/Valloric/ycmd/issues/67

The semantic engine is called either if the user enters a semantic trigger, or if they _force_ semantic completion. There may be many reasons why the semantic engine might return no results, but most commonly:

* there are errors in the file, and the semantic engine cannot parse it
* the semantic engine is not sophisticated enough (or it is not possible) to present a meaningful, correct or complete set of semantic suggestions (e.g. c++ temples, c macros, python duck-typing)
* the user's query string does not match any semantic completion suggestion

Previously, in all of the above cases, YCMD would return either no suggestions, or an error. This could make it seem like YCMD was not functioning correctly. Indeed in some cases (as discussed in the issues) this can lead to awkward workflow problems.

API Changes
==========

The `get_completions` request may now optionally contain a list `errors` containing the details of 0 or more errors raised by the semantic completion engine in addition to (a possibly empty) list of completion suggestions. The contents of each entry is identical to the `get_completions` response when no completion suggestions are returned, and only an exception is returned. That is:

```
EXCEPTION :=
{
    'exception' : {
        'TYPE': "<python exeption type name>"
    },
    'message': "<error description>",
    'traceback': "<stack trace of where the exception was thrown>"
}

COMPLETION :=
{
    'insertion_text': "completion text to be inserted",
    'extra_menu_info': "additional display information, e.g. type, [ID], [file], <snip>",
    'menu_text': "alternative text to display in the menu, in place of insertion_text",
    'detailed_info': "additional info for separate display, such as documentation/overloads, etc."
    'kind': "character defining type of suggestion",
    'extra_data': "completer-specific additional data. currently not documented",
}

COMPLETION_SUGGESTIONS_RESPONSE :=
{
    'completions': [ <potentially empty list of COMPLETION> ],
    'completion_start_column': column index of start point for completion suggestion,
    'errors': [ <potentially empty list of EXCEPTION> ]
}

EXCEPTION_RESPONSE := EXCEPTION

GET_COMPLETIONS_RESPONSE := EXCEPTION_RESPONSE | COMPLETION_SUGGESTIONS_RESPONSE
```

Rationale
=======

YCMD's `README.md` says:

>There are two main use-cases for code-completion:
>
>1. The user knows which name they're looking for, they just don't want to type
>   the whole name.
>2. The user either doesn't know the name they need or isn't sure what the name
>   is. This is also known as the "API exploration" use-case.
>
>The first use case is the most common one and is trivially addressed with the
>identifier completion engine (which BTW is blazing fast). The second one needs
>semantic completion.

This change is pretty much 100% targeted at the first use-case.

I think it is arguable that it is in keeping with YCMD's ethos of 'the user knows what they want most of the time' that in all of the above listed cases for no semantic suggestions, offering identifier completions is better than offering nothing.

This behaviour works well for dynamic languages, in particular those using 'duck typing' or similar semantics, such as python. This naturally extends to compiled languages using compile-time duck typing (c++ templates, c macros).

However, it is totally legitimate concern that users typing a semantic trigger expect a certain amount of accuracy from the completion suggestions (the 2nd use case in README.md). This is particularly important in the case where the query string does not match any of the semantic completions supplied by the semantic engine, and we filter them out. In order to help with this mental model, we tag all identifier completions with `[ID]` in the completion menu. There is then a subtle, but clear indication that the suggestion is _not_ from the semantic engine. Existing "general" completer already tag their suggestions with `[File]`, `[Dir]` and `[snip]` etc.

What this change does
==================

When the semantic engine is called after a trigger (*not* after forced completion), and either no completion suggestions are returned or the completer throws an exception, we query the "general" completer (typically the identifier completer) and return those suggestions instead. 

In order to make this more obvious to the user, we mark suggestions from the
identifier completer with [ID] in completions menu. 

In order to preserve the behaviour of showing exceptions thrown by the completer to the user (e.g. in case they have an error in their `.ycm_extra_conf.py`, we return the exceptions as well as the identifier completions and print this exception to the log file.

Contentious issues
===============

As discussed on the above tracker issues, the general behaviour is somewhat divisive. I personally, having been using it for some time and *really* like it. I'm almost certainly going to maintain it in my fork. I thought about adding a switch (on by default), but thought I would avoid that because more switches is a _bad thing_(TM).

Also, this is a change in the YCMD API and as such requires minor changes in clients to display these additional errors. The `README.md` says that there is a SemVer API version, but i wan't able to find one other than the `CORE_VERSION`

### CODE REVIEW

**Comment by Valloric** [comment]
I'm fine with the idea of this change (haven't looked at the implementation yet).

> I thought about adding a switch (on by default), but thought I would avoid that because more switches is a bad thing(TM).

Your instinct to _not_ add Yet Another User-configurable Option was indeed correct. :)

> Also, this is a change in the YCMD API and as such requires minor changes in clients to display these additional errors.

Yes. It would be great if you could summarize such API changes here so that @abingham, @ptrv and others can quickly see what's changing.

> The README.md says that there is a SemVer API version, but i wan't able to find one other than the CORE_VERSION

We're still in the 0.x phase where anything goes. Still, we try not to break people.

P.S. I spell it as "ycmd", not "YCMD". So YCM is all-caps (since it's an acronym), but ycmd is lowercase (since it's a full name).

File: ycmd/handlers.py [hunk 1]
```diff
@@ -89,17 +90,39 @@ def RunCompleterCommand():
...
-  filetypes = request_data[ 'filetypes' ]
-  completer = ( _server_state.GetFiletypeCompleter( filetypes ) if
-                do_filetype_completion else
-                _server_state.GetGeneralCompleter() )
 
-  return _JsonResponse( BuildCompletionResponse(
-      completer.ComputeCandidates( request_data ),
-      request_data.CompletionStartColumn() ) )
+  general_completer = _server_state.GetGeneralCompleter()
+  kw = {}
```
Comment (Valloric) [inline ycmd/handlers.py]: `kw` is a bit short. I can't tell at-a-glance what this variable is about which is a solid hint that it should be longer.

File: ycmd/handlers.py [hunk 2]
```diff
@@ -89,17 +90,39 @@ def RunCompleterCommand():
...
+    try:
+      completions = ( _server_state.GetFiletypeCompleter(
+                                  request_data[ 'filetypes' ] )
+                                 .ComputeCandidates( request_data ) )
+
+      if not completions and not forced_filetype_completion:
+        completions = general_completer.ComputeCandidates( request_data )
+    except Exception as e:
+      if not forced_filetype_completion:
+        completions = general_completer.ComputeCandidates( request_data )
```
Comment (Valloric) [inline ycmd/handlers.py]: This repeats the code from above. I think you can restructure this to avoid the repetition.

Maybe something like:
```
completions = None
if do_filetype_completion:
  try:
    completions = <get semantic>
  except:
    if forced:
      raise
    else:
      <store error>
if not completions and not forced:
  completions = <get ident>
return completions + error data if any
```

File: ycmd/handlers.py [hunk 3]
```diff
@@ -89,17 +90,39 @@ def RunCompleterCommand():
...
+        kw[ 'errors' ] = [ BuildExceptionResponse( e, tb ) ]
+      else:
+        raise
+  else:
+    completions = general_completer.ComputeCandidates( request_data )
+
+  return _JsonResponse(
+      BuildCompletionResponse( completions,
+                               request_data.CompletionStartColumn(),
+                               **kw ) )
```
Comment (Valloric) [inline ycmd/handlers.py]: With the pseudo-code I posted above, you could probably also get rid of the whole `**kw` thing and just do `errors = errors` here. Should be easier for readers to understand.

File: ycmd/server_state.py [hunk 4]
```diff
@@ -101,10 +101,12 @@ def ShouldUseGeneralCompleter( self, request_data ):
   def ShouldUseFiletypeCompleter( self, request_data ):
```
Comment (Valloric) [inline ycmd/server_state.py]: Consider adding a comment here noting that a (bool, bool) is returned and what each of them is.

File: ycmd/tests/get_completions_test.py [hunk 5]
```diff
@@ -330,9 +1129,17 @@ def GetCompletions_CsCompleter_NonForcedReturnsNoResults_test():
                                   column_num = 21,
                                   force_semantic = False,
                                   query = 'Date' )
-  results = app.post_json( '/completions', completion_data ).json[ 'completions' ]
-
-  assert_that( results, empty() )
+  results = app.post_json( '/completions', completion_data ).json
+
+  # there are no semantic completions. However, we fall back to identifier
+  # completer in this case.
```
Comment (Valloric) [inline ycmd/tests/get_completions_test.py]: It's probably a better idea to not mix the fallback to identifiers into this. Use forced completion so that you can ignore it.

File: ycmd/tests/testdata/general_fallback/lang_c.c [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+static void do_something( int );
+static void do_another_thing( int );
+
+// TESTCASE1: use of . on macro parameter. Macros used this way are compile-time
+// duck-typing. Note this in this particular instance, libclang actually
+// offers semantic completions if there is only call to this macro
+// (e.g. DO_SOMETHING_TO( a_test ) - clever little shrew), so we don't call it
+
+/*       1         2         3         4
+1234567890123456789012345678901234567890123456789 */
```
Comment (Valloric) [inline ycmd/tests/testdata/general_fallback/lang_c.c]: What's the purpose of this list of numbers? Is it to facilitate seeing what column is where? If that's the case, readers should configure their editor to show them a column number in the status line. :) The numbers are kinda distracting.

File: ycmd/tests/get_completions_test.py [hunk 7]
```diff
@@ -115,9 +123,800 @@ def GetCompletions_IdentifierCompleter_WorksForSpecialIdentifierChars_test():
...
+  Tests for:
+    - [X] multiple languages: (use cases)
+      - [X] C:      macros
+      - [X] C++:    templates
+      - [X] Python: properties set from external method
+    - [X] Forced vs. Triggered semantic completion
+    - [X] Identifiers tagged with '[ID]'
+    - [X] Filter/sort based on query
+    - [X] Min chars for completion (only applies to ID)
+    - [X] Extra conf throws exception
```
Comment (Valloric) [inline ycmd/tests/get_completions_test.py]: I have a feeling you're testing too many things at once.

For instance, I'd have just one test that tests the fallback (for say C++). The codepath that implements the fallback is agnostic of the semantic engine, so testing that the fallback works for various semantic engines seems overkill to me.

Likewise, for semantic engine tests, just use force_semantic to not have to test the fallback logic. That's tested in the fallback test.

A good design idea for tests is that if one logical thing fails in the code, one test should fail. That's often not achievable, but something we should strive for. The idea is that if one logical component fails and 100 tests fail because of it, tracking down exactly which component failed because non-trivial.

File: ycmd/tests/testdata/general_fallback/lang_c.c (hunk 6 above)
Comment (puremourning) [inline ycmd/tests/testdata/general_fallback/lang_c.c]: Yeah, it's a ruler. I find reading the tests and the associated code quite a painfully slow and cumbersome when you have to read 2 numbers and then navigate to that exact spot in the file.

You can get vim to show column numbers, of course, but you still have to pretty much put the cursor there. The use case for the ruler is a vertical split where i'm writing (or reading) the test in one split and i can just visually see (via set nu and this ruler) the exact spot where the test needs to trigger (or is already triggering).

I'm happy to remove them if you feel strongly about it. Of course removing them changes all the offsets which is a bit tedious, but hey ho :)

File: ycmd/tests/get_completions_test.py (hunk 7 above)
Comment (puremourning) [inline ycmd/tests/get_completions_test.py]: Interesting perspective. Happy to change this.

FWIW my perspective has always been that it's really useful to write regression tests (in particular integration tests testing an interface - as these are) from the functional perspective, so that if the code is changed (arbitrarily) then the tests should still work. I've always extended that to mean "even if the underlying codepath is similar, then the functional test still stands". i.e. if we made a change to the python completer which somehow interfaced badly with this, then we'd catch it.

Like I said, happy to change this.

**Comment by puremourning** [comment]
Thanks for the review, @Valloric. I'm back at work now, so might be a few days before i can finish it up.

**Comment by puremourning** [comment]
> P.S. I spell it as "ycmd", not "YCMD". So YCM is all-caps (since it's an acronym), but ycmd is lowercase (since it's a full name).

Yeah, i know, but safari has a habit of auto-"correcting" ycmd to something else, so i've got in the habit of typing it in upper case. I suppose I should find a way of telling it ycmd is a real word instead :p

File: ycmd/tests/testdata/general_fallback/lang_c.c (hunk 6 above)
Comment (Valloric) [inline ycmd/tests/testdata/general_fallback/lang_c.c]: I'm not crazy enough to feel strongly about them. Feel free to leave them in. :)

File: ycmd/tests/get_completions_test.py (hunk 7 above)
Comment (Valloric) [inline ycmd/tests/get_completions_test.py]: That perspective is exemplary but in my experience leads to tests that can be hard to maintain and a general test suite that isn't as helpful as it could be. I've worked in a codebase with tons of tests all testing the same thing indirectly, and when one such important component failed, it was a nightmare to figure out what actually went wrong. You'd know _something_ went south because 100+ tests failed, but good luck figuring it out.

Since then I've focused on the mantra that "all functionality should be tested while trying to minimize test path duplication."

Think of it like a compiler error: if you forget a semicolon, would you rather see a spew of 100+ errors or just one saying "hey you forgot a semicolon here"? C++ for instance has compilers that spew out an ungodly mess of text on incorrect template usage and figuring out what exactly went wrong is hard.

> my perspective has always been that it's really useful to write regression tests (in particular integration tests testing an interface - as these are) from the functional perspective, so that if the code is changed (arbitrarily) then the tests should still work.

Black-box testing can be very useful, but isolated unit-tests IMO are more useful. But the principle you're espousing of "tests should still work if the code is changed" is absolutely something that's critical and IMO not at odds with my approach.

**Comment by Valloric** [comment]
> Thanks for the review, @Valloric. I'm back at work now, so might be a few days before i can finish it up.

How _dare you_ sir have a life outside of open-source! :D

Seriously, take all the time you need. We're not in a rush.

File: ycmd/tests/get_completions_test.py (hunk 5 above)
Comment (puremourning) [inline ycmd/tests/get_completions_test.py]: Hmmm. The following tests seem to be explicitly testing "non-forced" semantic completions (or the interaction of forced and non-forced in the latter case). So i'm tempted to leave them as-is, so as not to interfere with the idea behind the original test.

```python
def GetCompletions_CsCompleter_NonForcedReturnsNoResults_test():
def GetCompletions_CsCompleter_ForcedDividesCache_test():
```

These ones test exceptions from the completer, and in particular the common-ish case where there is no .ycm_extra_conf (and therefore not compatible with my generic `GetCompletions_RunTest` function), but i think similar tests do exist, so i've reverted the changes:

```python
GetCompletions_ClangCompleter_ExceptionWhenNoFlagsFromExtraConf_test
GetCompletions_ClangCompleter_ExceptionWhenNoFlagsFromExtraConf_ID_test
```

What do you think?

File: ycmd/tests/get_completions_test.py (hunk 7 above)
Comment (puremourning) [inline ycmd/tests/get_completions_test.py]: I’m going to trim down the tests to just:

- one language (clang completer-based)
- forced vs triggered semantic (fallback only in 2nd case)
- exception thrown, vs. just no completions for query
- min chars for identifier completion

Would anyone object to keeping the testdata files as they are? This is not just laziness, but I do find it useful to open the testsdata/* files to do manual testing of a particular feature (almost as a demo). And the test files i wrote mostly for manual testing in the first instance to ensure the feature feels right and works sensibly.

Comment (puremourning) [inline ycmd/tests/get_completions_test.py]: Done. I've trimmed the tests to just one each covering each relevant (IMO) code path.

File: ycmd/server_state.py (hunk 4 above)
Comment (puremourning) [inline ycmd/server_state.py]: Added method docstring.

File: ycmd/handlers.py (hunk 3 above)
Comment (puremourning) [inline ycmd/handlers.py]: Yep done.

File: ycmd/handlers.py (hunk 2 above)
Comment (puremourning) [inline ycmd/handlers.py]: done.

File: ycmd/handlers.py (hunk 1 above)
Comment (puremourning) [inline ycmd/handlers.py]: done.

**Comment by puremourning** [comment]
> Yes. It would be great if you could summarize such API changes here so that @abingham, @ptrv and others can quickly see what's changing.

Done above. I'm not aware of any "official" ycmd interface docs, so i wrote a (very) quick spec. of the `get_completions` response above, with a summary of the changes. I suppose we should add "document the API" to the list of things to do in the future :)

I've pushed new commit addressing (most of) your comments.

Thanks,
ben

File: ycmd/tests/get_completions_test.py (hunk 5 above)
Comment (Valloric) [inline ycmd/tests/get_completions_test.py]: You've been looking at these tests more recently than I have and you make good points; I'll leave the decision of what tests should be changed up to you. I trust your judgment.

**Comment by Valloric** [comment]
> I suppose we should add "document the API" to the list of things to do in the future :)

Ha! Right you are. I've been "meaning to do that" for quite a while now. :D

From my end, this is ready to merge. I'm not sure did you want to do something else on it as well before I merged it, so ping back if this is ready.

**Comment by puremourning** [comment]
@Valloric yep this is ready to merge.

**Comment by Valloric** [comment]
@homu r+
